src/OrchardCore.Modules/OrchardCore.Media.Azure/Services/MediaBlobFileProvider.cs

```diff
@@ -0,0 +1,93 @@
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.FileProviders;
+using Microsoft.Extensions.FileProviders.Physical;
+using Microsoft.Extensions.Logging;
+using OrchardCore.FileStorage;
+
+namespace OrchardCore.Media.Azure.Services
+{
+    public class MediaBlobFileProvider : PhysicalFileProvider, IMediaFileProvider, IMediaCacheFileProvider, IMediaFileStoreCache
+    {
+        // Use default stream copy buffer size to stay in gen0 garbage collection;
+        private const int StreamCopyBufferSize = 81920;
+
+        private readonly ILogger<MediaBlobFileProvider> _logger;
+
+        public MediaBlobFileProvider(ILogger<MediaBlobFileProvider> logger, PathString virtualPathBase, string root) : base(root)
+        {
+            _logger = logger;
+            VirtualPathBase = virtualPathBase;
+        }
+
+        public MediaBlobFileProvider(ILogger<MediaBlobFileProvider> logger, PathString virtualPathBase, string root, ExclusionFilters filters) : base(root, filters)
+        {
+            _logger = logger;
+            VirtualPathBase = virtualPathBase;
+        }
+
+        public PathString VirtualPathBase { get; }
+
+        public Task<bool> IsCachedAsync(string path)
+        {
+            // Opportunity here to save metadata and/or provide cache validation / integrity checks.
+
+            var fileInfo = GetFileInfo(path);
```

Comment:
The caching side of this file provider is abstracted now, but the fileprovider side is still an `IFileProvider` because that's what the `StaticFileMiddleware` needs to recieve. So half an abstraction :)